File: PhoneticTokenizer/phonetok/phonotactics.py
"""Phonotactics: which sounds go well together, and how to build a word out of them.

Two kinds of knowledge, one object:

* **the law** - what English lets a syllable be made of: the onset list and
  the sonority of the coda (:mod:`phonetok.syllables`).  :meth:`Phonotactics.well_formed`
  is that test, and it needs no data.
* **the habit** - how often one sound actually follows another, counted over
  every pronunciation the lexicon holds (or any pronunciations handed to
  :meth:`Phonotactics.fit`).  :meth:`affinity` is the pointwise mutual
  information of a pair, in bits: ``0`` is chance, positive is *these go
  together*, negative is *these avoid each other*.  :meth:`score` is the mean
  log-probability of a whole word's transitions, so *strength* scores well and
  *ngstrelk* does not.

:meth:`build` uses both to coin new words - an onset, a vowel and a coda per
syllable, each sampled by how often it follows what came before - and
:meth:`blend` finds the best joint for a portmanteau.  Counts, not a model:
:meth:`to_dict` writes them out and :meth:`from_dict` reads them back.
"""

from __future__ import annotations

import math
import random
from collections import Counter
from collections.abc import Iterable, Sequence

from .phones import BOUNDARY, CONSONANTS, VOWELS, base, is_vowel, strip_stress, with_stress
from .syllables import ONSETS, Syllable, is_legal_coda, is_legal_onset, syllabify

Phones = list[str]


def _sample(counter: Counter, rng: random.Random, exclude: set | None = None):
    """A weighted draw, in key order so that the same seed draws the same thing on every machine."""
    items = sorted((k, c) for k, c in counter.items() if not exclude or k not in exclude)
    if not items:
        return None
    total = sum(c for _, c in items)
    r = rng.random() * total
    acc = 0.0
    for k, c in items:
        acc += c
        if r < acc:
            return k
    return items[-1][0]


class Phonotactics:
    """Sound-to-sound habits counted from pronunciations, plus the law of the syllable."""

    def __init__(self, stress: bool = False, smoothing: float = 0.5) -> None:
        self.stress = stress
        """Whether the counts keep the vowels' stress digits (sparser, but truer to speech)."""
        self.smoothing = smoothing
        self.bigrams: Counter[tuple[str, str]] = Counter()
        self.unigrams: Counter[str] = Counter()
        self.followers: dict[str, Counter[str]] = {}
        self.onsets_initial: Counter[tuple[str, ...]] = Counter()
        self.onsets_medial: Counter[tuple[str, ...]] = Counter()
        self.nuclei: Counter[str] = Counter()
        self.codas_medial: Counter[tuple[str, ...]] = Counter()
        self.codas_final: Counter[tuple[str, ...]] = Counter()
        self.syllable_counts: Counter[int] = Counter()
        self.words = 0

    # -- fitting -------------------------------------------------------------

    def _key(self, phone: str) -> str:
        return phone if self.stress else base(phone)

    def fit(self, pronunciations: Iterable[Sequence[str]]) -> "Phonotactics":
        """Count the transitions and the syllable parts of every pronunciation.  Returns ``self``."""
        for phones in pronunciations:
            phones = list(phones)
            if not phones:
                continue
            keys = [BOUNDARY] + [self._key(p) for p in phones] + [BOUNDARY]
            for a, b in zip(keys, keys[1:]):
                self.bigrams[(a, b)] += 1
                self.unigrams[a] += 1
                self.followers.setdefault(a, Counter())[b] += 1
            self.unigrams[BOUNDARY] += 1
            syllables = syllabify(phones)
            self.syllable_counts[len(syllables)] += 1
            for i, s in enumerate(syllables):
                (self.onsets_initial if i == 0 else self.onsets_medial)[s.onset] += 1
                self.nuclei[s.nucleus if self.stress else with_stress(s.nucleus, 0) if is_vowel(s.nucleus) else s.nucleus] += 1
                (self.codas_final if i == len(syllables) - 1 else self.codas_medial)[s.coda] += 1
            self.words += 1
        return self

    @classmethod
    def from_lexicon(cls, lexicon, stress: bool = False) -> "Phonotactics":
        """Fit on every pronunciation of a :class:`phonetok.lexicon.Lexicon`."""
        return cls(stress=stress).fit(phones for _, phones in lexicon.items())

    @property
    def alphabet_size(self) -> int:
        return len(self.unigrams) or 1

    # -- the habit -----------------------------------------------------------

    def prob(self, a: str, b: str) -> float:
        """``P(b | a)``, add-k smoothed over the alphabet seen."""
        a, b = self._key(a), self._key(b)
        k = self.smoothing
        return (self.bigrams[(a, b)] + k) / (self.unigrams[a] + k * self.alphabet_size)

    def log_prob(self, a: str, b: str) -> float:
        return math.log(self.prob(a, b))

    def affinity(self, a: str, b: str) -> float:
        """How well ``b`` follows ``a``, as pointwise mutual information in bits (0 = chance).

        ``affinity("S", "T")`` is positive (the commonest cluster in the language);
        ``affinity("T", "L")`` is negative (no English syllable starts with *tl*).
        """
        a, b = self._key(a), self._key(b)
        total = sum(self.unigrams.values()) or 1
        k = self.smoothing
        p_ab = (self.bigrams[(a, b)] + k) / (total + k * self.alphabet_size ** 2)
        p_a = (self.unigrams[a] + k) / (total + k * self.alphabet_size)
        p_b = (self.unigrams[b] + k) / (total + k * self.alphabet_size)
        return math.log2(p_ab / (p_a * p_b))

    def transitions(self, phones: Sequence[str]) -> list[tuple[str, str, float]]:
        """Every transition of a word with its log-probability, the boundaries included."""
        keys = [BOUNDARY] + [self._key(p) for p in phones] + [BOUNDARY]
        return [(a, b, self.log_prob(a, b)) for a, b in zip(keys, keys[1:])]

    def score(self, phones: Sequence[str]) -> float:
        """The mean log-probability per transition: how English these sounds are in this order."""
        steps = self.transitions(phones)
        if not steps:
            return 0.0
        return sum(lp for _, _, lp in steps) / len(steps)

    def weakest(self, phones: Sequence[str]) -> tuple[int, float]:
        """The index (into the transitions) and log-probability of the least likely step."""
        steps = self.transitions(phones)
        i = min(range(len(steps)), key=lambda i: steps[i][2])
        return i, steps[i][2]

    def next(self, phone: str, k: int = 5) -> list[tuple[str, float]]:
        """The ``k`` likeliest sounds after ``phone`` (``"#"`` for the start of a word), with ``P``."""
        a = self._key(phone)
        table = self.followers.get(a)
        if not table:
            return []
        total = sum(table.values())
        return [(b, c / total) for b, c in table.most_common(k)]

    # -- the law -------------------------------------------------------------

    @staticmethod
    def well_formed(phones: Sequence[str]) -> bool:
        """Could an English speaker say this: every onset legal, every coda falling, a vowel somewhere?"""
        phones = list(phones)
        if not phones or not any(is_vowel(p) for p in phones):
            return False
        for s in syllabify(phones):
            if not is_legal_onset(s.onset) or not is_legal_coda(s.coda):
                return False
        return True

    @staticmethod
    def violations(phones: Sequence[str]) -> list[str]:
        """What is wrong with a sequence, in words; ``[]`` when nothing is."""
        phones = list(phones)
        if not phones:
            return ["empty"]
        out: list[str] = []
        if not any(is_vowel(p) for p in phones):
            out.append("no vowel")
        for s in syllabify(phones):
            if not is_legal_onset(s.onset):
                out.append(f"{' '.join(s.onset)} cannot start a syllable")
            if not is_legal_coda(s.coda):
                out.append(f"{' '.join(s.coda)} cannot end a syllable")
        return out

    # -- building words ------------------------------------------------------

    def build(
        self, rng: random.Random | int | None = None, syllables: int | None = None,
        min_syllables: int = 1, max_syllables: int = 3, avoid: Iterable[Sequence[str]] = (), tries: int = 50,
    ) -> Phones:
        """Coin a pronounceable word from the habits: one onset, vowel and coda per syllable.

        Each part is drawn by how often it was seen in that position, and the joints
        between parts are weighted by :meth:`prob` so the sounds that meet are sounds that
        go together.  One syllable carries the primary stress.  ``avoid`` is a set of
        pronunciations not to return (the lexicon's, to make the word new).
        """
        if not isinstance(rng, random.Random):
            rng = random.Random(rng)
        if not self.words:
            raise ValueError("fit the phonotactics on some pronunciations first")
        forbidden = {tuple(strip_stress(p)) for p in avoid}
        best: Phones = []
        for _ in range(tries):
            n = syllables if syllables else _sample_syllables(self, rng, min_syllables, max_syllables)
            stressed = 0 if n == 1 else (0 if rng.random() < 0.6 else rng.randrange(n))
            phones: Phones = []
            ok = True
            for i in range(n):
                prev = phones[-1] if phones else BOUNDARY
                onset = self._draw(self.onsets_initial if i == 0 else self.onsets_medial, rng, prev)
                if onset is None:
                    ok = False
                    break
                phones.extend(onset)
                prev = phones[-1] if phones else BOUNDARY
                nucleus = self._draw_nucleus(rng, prev, 1 if i == stressed else 0)
                phones.append(nucleus)
                last = i == n - 1
                coda = self._draw(self.codas_final if last else self.codas_medial, rng, nucleus)
                if coda is None:
                    ok = False
                    break
                phones.extend(coda)
            if not ok or not self.well_formed(phones):
                continue
            if tuple(strip_stress(phones)) in forbidden:
                continue
            return phones
        return best

    def _draw(self, table: Counter, rng: random.Random, prev: str):
        """A syllable part from ``table``, weighted by its count and by how well it follows ``prev``."""
        if not table:
            return ()
        weighted: Counter = Counter()
        for part, count in table.items():
            first = part[0] if part else None
            weight = count * (self.prob(prev, first) * self.alphabet_size if first else 1.0)
            weighted[part] = weight
        return _sample(weighted, rng)

    def _draw_nucleus(self, rng: random.Random, prev: str, stress: int) -> str:
        weighted: Counter = Counter()
        for nucleus, count in self.nuclei.items():
            if not is_vowel(nucleus):
                continue
            weighted[nucleus] = count * self.prob(prev, nucleus) * self.alphabet_size
        chosen = _sample(weighted, rng) or "AH0"
        return with_stress(chosen, stress)

    def blend(self, a: Sequence[str], b: Sequence[str]) -> tuple[Phones, int, int]:
        """The best portmanteau of two words: ``(phones, syllables kept of a, syllables dropped of b)``.

        Every cut that keeps at least one syllable of each is scored by how well the
        sounds at the joint go together and how close to half of each word survives;
        *breakfast* + *lunch* gives *brunch*-like joints, *smoke* + *fog* gives *smog*.
        """
        if not any(is_vowel(p) for p in a) or not any(is_vowel(p) for p in b):
            raise ValueError("both words need a vowel to be blended")
        sa, sb = syllabify(list(a)), syllabify(list(b))
        best: tuple[float, Phones, int, int] | None = None
        candidates: list[tuple[Phones, int, int]] = []
        # joints at syllable edges, and inside the syllables at the vowel
        for i in range(1, len(sa) + 1):
            for j in range(0, len(sb)):
                head = [p for s in sa[:i] for p in s.phones]
                tail = [p for s in sb[j:] for p in s.phones]
                candidates.append((head + tail, i, j))
                # keep a's onset (and vowel) and take b's rime
                head_open = [p for s in sa[: i - 1] for p in s.phones] + list(sa[i - 1].onset)
                candidates.append((head_open + list(sb[j].rime) + [p for s in sb[j + 1:] for p in s.phones], i, j))
        for phones, i, j in candidates:
            if not self.well_formed(phones) or len(phones) < 2:
                continue
            joint = self.score(phones)
            balance = -abs((i / len(sa)) - 0.5) - abs(((len(sb) - j) / len(sb)) - 0.5)
            length = len(phones) / max(len(a), len(b))  # a blend that keeps more of its parents reads as both
            total = joint + 0.5 * balance + 0.2 * length
            if best is None or total > best[0]:
                best = (total, phones, i, j)
        if best is None:
            return list(a) + list(b), len(sa), 0
        phones = best[1]
        # one primary stress
        if sum(1 for p in phones if p.endswith("1")) > 1:
            seen = False
            for k, p in enumerate(phones):
                if p.endswith("1"):
                    if seen:
                        phones[k] = with_stress(p, 2)
                    seen = True
        if not any(p.endswith("1") for p in phones):
            for k, p in enumerate(phones):
                if is_vowel(p):
                    phones[k] = with_stress(p, 1)
                    break
        return phones, best[2], best[3]

    # -- persistence ---------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "stress": self.stress, "smoothing": self.smoothing, "words": self.words,
            "bigrams": [[a, b, c] for (a, b), c in sorted(self.bigrams.items())],
            "onsets_initial": [[" ".join(k), c] for k, c in sorted(self.onsets_initial.items())],
            "onsets_medial": [[" ".join(k), c] for k, c in sorted(self.onsets_medial.items())],
            "nuclei": sorted(self.nuclei.items()),
            "codas_medial": [[" ".join(k), c] for k, c in sorted(self.codas_medial.items())],
            "codas_final": [[" ".join(k), c] for k, c in sorted(self.codas_final.items())],
            "syllable_counts": sorted(self.syllable_counts.items()),
        }

    @classmethod
    def from_dict(cls, d: dict) -> "Phonotactics":
        p = cls(stress=bool(d.get("stress", False)), smoothing=float(d.get("smoothing", 0.5)))
        p.words = int(d.get("words", 0))
        for a, b, c in d.get("bigrams", []):
            p.bigrams[(a, b)] = c
            p.unigrams[a] += c
            p.followers.setdefault(a, Counter())[b] += c
        p.unigrams[BOUNDARY] += p.words  # the final boundary of every word
        for name in ("onsets_initial", "onsets_medial", "codas_medial", "codas_final"):
            table = getattr(p, name)
            for k, c in d.get(name, []):
                table[tuple(k.split()) if k else ()] = c
        for k, c in d.get("nuclei", []):
            p.nuclei[k] = c
        for k, c in d.get("syllable_counts", []):
            p.syllable_counts[int(k)] = c
        return p

    def __repr__(self) -> str:
        return f"Phonotactics(words={self.words}, pairs={len(self.bigrams)}, stress={self.stress})"


def _sample_syllables(p: Phonotactics, rng: random.Random, lo: int, hi: int) -> int:
    table = Counter({n: c for n, c in p.syllable_counts.items() if lo <= n <= hi})
    return _sample(table, rng) or lo

File: PhoneticTokenizer/go/phonetok/phonotactics.go
package phonetok

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// Phonotactics (the port of phonotactics.py): which sounds go well together,
// and how to build a word out of them.

// Phonotactics are sound-to-sound habits counted from pronunciations, plus the law of the syllable.
type Phonotactics struct {
	Stress         bool
	Smoothing      float64
	Bigrams        map[[2]string]int
	Unigrams       map[string]int
	followers      map[string]map[string]int
	followerOrder  map[string][]string
	OnsetsInitial  map[string]int // clusters joined by spaces
	OnsetsMedial   map[string]int
	Nuclei         map[string]int
	CodasMedial    map[string]int
	CodasFinal     map[string]int
	SyllableCounts map[int]int
	Words          int
}

// NewPhonotactics is an empty table.
func NewPhonotactics(stress bool) *Phonotactics {
	return &Phonotactics{
		Stress: stress, Smoothing: 0.5, Bigrams: map[[2]string]int{}, Unigrams: map[string]int{},
		followers: map[string]map[string]int{}, followerOrder: map[string][]string{},
		OnsetsInitial: map[string]int{}, OnsetsMedial: map[string]int{}, Nuclei: map[string]int{},
		CodasMedial: map[string]int{}, CodasFinal: map[string]int{}, SyllableCounts: map[int]int{},
	}
}

func (p *Phonotactics) key(phone string) string {
	if p.Stress {
		return phone
	}
	return Base(phone)
}

// Fit counts the transitions and the syllable parts of every pronunciation.
func (p *Phonotactics) Fit(pronunciations [][]string) *Phonotactics {
	for _, phones := range pronunciations {
		if len(phones) == 0 {
			continue
		}
		keys := make([]string, 0, len(phones)+2)
		keys = append(keys, Boundary)
		for _, ph := range phones {
			keys = append(keys, p.key(ph))
		}
		keys = append(keys, Boundary)
		for i := 0; i+1 < len(keys); i++ {
			a, b := keys[i], keys[i+1]
			p.Bigrams[[2]string{a, b}]++
			p.Unigrams[a]++
			if p.followers[a] == nil {
				p.followers[a] = map[string]int{}
			}
			if _, seen := p.followers[a][b]; !seen {
				p.followerOrder[a] = append(p.followerOrder[a], b)
			}
			p.followers[a][b]++
		}
		p.Unigrams[Boundary]++
		syllables := Syllabify(phones, true)
		p.SyllableCounts[len(syllables)]++
		for i, s := range syllables {
			if i == 0 {
				p.OnsetsInitial[strings.Join(s.Onset, " ")]++
			} else {
				p.OnsetsMedial[strings.Join(s.Onset, " ")]++
			}
			nucleus := s.Nucleus
			if !p.Stress && IsVowel(nucleus) {
				nucleus = WithStress(nucleus, 0)
			}
			p.Nuclei[nucleus]++
			if i == len(syllables)-1 {
				p.CodasFinal[strings.Join(s.Coda, " ")]++
			} else {
				p.CodasMedial[strings.Join(s.Coda, " ")]++
			}
		}
		p.Words++
	}
	return p
}

// PhonotacticsFromLexicon fits on every pronunciation of a lexicon.
func PhonotacticsFromLexicon(lex *Lexicon, stress bool) *Phonotactics {
	items := lex.Items()
	prons := make([][]string, len(items))
	for i, e := range items {
		prons[i] = e.Phones
	}
	return NewPhonotactics(stress).Fit(prons)
}

// AlphabetSize is how many distinct first symbols were seen (at least 1).
func (p *Phonotactics) AlphabetSize() int {
	if len(p.Unigrams) == 0 {
		return 1
	}
	return len(p.Unigrams)
}

// Prob is P(b | a), add-k smoothed over the alphabet seen.
func (p *Phonotactics) Prob(a, b string) float64 {
	a, b = p.key(a), p.key(b)
	k := p.Smoothing
	return (float64(p.Bigrams[[2]string{a, b}]) + k) / (float64(p.Unigrams[a]) + k*float64(p.AlphabetSize()))
}

// LogProb is ln P(b | a).
func (p *Phonotactics) LogProb(a, b string) float64 { return math.Log(p.Prob(a, b)) }

// Affinity is how well b follows a, as pointwise mutual information in bits (0 = chance).
func (p *Phonotactics) Affinity(a, b string) float64 {
	a, b = p.key(a), p.key(b)
	total := 0
	for _, c := range p.Unigrams {
		total += c
	}
	if total == 0 {
		total = 1
	}
	k := p.Smoothing
	A := float64(p.AlphabetSize())
	pab := (float64(p.Bigrams[[2]string{a, b}]) + k) / (float64(total) + k*A*A)
	pa := (float64(p.Unigrams[a]) + k) / (float64(total) + k*A)
	pb := (float64(p.Unigrams[b]) + k) / (float64(total) + k*A)
	return math.Log2(pab / (pa * pb))
}

// Transition is one step of a word with its log-probability.
type Transition struct {
	From, To string
	LogProb  float64
}

// Transitions are every transition of a word, the boundaries included.
func (p *Phonotactics) Transitions(phones []string) []Transition {
	keys := make([]string, 0, len(phones)+2)
	keys = append(keys, Boundary)
	for _, ph := range phones {
		keys = append(keys, p.key(ph))
	}
	keys = append(keys, Boundary)
	out := make([]Transition, 0, len(keys)-1)
	for i := 0; i+1 < len(keys); i++ {
		out = append(out, Transition{keys[i], keys[i+1], p.LogProb(keys[i], keys[i+1])})
	}
	return out
}

// Score is the mean log-probability per transition.
func (p *Phonotactics) Score(phones []string) float64 {
	steps := p.Transitions(phones)
	if len(steps) == 0 {
		return 0
	}
	sum := 0.0
	for _, s := range steps {
		sum += s.LogProb
	}
	return sum / float64(len(steps))
}

// Weakest is the index and log-probability of the least likely step.
func (p *Phonotactics) Weakest(phones []string) (int, float64) {
	steps := p.Transitions(phones)
	best := 0
	for i, s := range steps {
		if s.LogProb < steps[best].LogProb {
			best = i
		}
	}
	return best, steps[best].LogProb
}

// Follower is one likely next sound with its probability.
type Follower struct {
	Phone string
	P     float64
}

// Next is the k likeliest sounds after phone ("#" for the start of a word).
func (p *Phonotactics) Next(phone string, k int) []Follower {
	a := p.key(phone)
	table := p.followers[a]
	if len(table) == 0 {
		return nil
	}
	order := append([]string(nil), p.followerOrder[a]...)
	sort.SliceStable(order, func(i, j int) bool { return table[order[i]] > table[order[j]] })
	total := 0
	for _, c := range table {
		total += c
	}
	if k > len(order) {
		k = len(order)
	}
	out := make([]Follower, 0, k)
	for _, b := range order[:k] {
		out = append(out, Follower{b, float64(table[b]) / float64(total)})
	}
	return out
}

// WellFormed reports whether an English speaker could say this: every onset legal, every coda
// falling, a vowel somewhere.
func WellFormed(phones []string) bool {
	if len(phones) == 0 {
		return false
	}
	hasVowel := false
	for _, p := range phones {
		if IsVowel(p) {
			hasVowel = true
		}
	}
	if !hasVowel {
		return false
	}
	for _, s := range Syllabify(phones, true) {
		if !IsLegalOnset(s.Onset) || !IsLegalCoda(s.Coda) {
			return false
		}
	}
	return true
}

// Violations say what is wrong with a sequence; none when nothing is.
func Violations(phones []string) []string {
	if len(phones) == 0 {
		return []string{"empty"}
	}
	var out []string
	hasVowel := false
	for _, p := range phones {
		if IsVowel(p) {
			hasVowel = true
		}
	}
	if !hasVowel {
		out = append(out, "no vowel")
	}
	for _, s := range Syllabify(phones, true) {
		if !IsLegalOnset(s.Onset) {
			out = append(out, strings.Join(s.Onset, " ")+" cannot start a syllable")
		}
		if !IsLegalCoda(s.Coda) {
			out = append(out, strings.Join(s.Coda, " ")+" cannot end a syllable")
		}
	}
	return out
}

// -- building words --------------------------------------------------------------

// weighted is a table of parts with float weights, drawn in key order like Python's sorted().
type weightedItem struct {
	key    string // the part, phones joined by spaces (a tuple in Python)
	weight float64
}

func lessCluster(a, b string) bool {
	// Python compares tuples of strings element by element, a shorter prefix first
	pa, pb := clusterParts(a), clusterParts(b)
	for i := 0; i < len(pa) && i < len(pb); i++ {
		if pa[i] != pb[i] {
			return pa[i] < pb[i]
		}
	}
	return len(pa) < len(pb)
}

func clusterParts(s string) []string {
	if s == "" {
		return nil
	}
	return strings.Split(s, " ")
}

func sample(items []weightedItem, rng Rng) (string, bool) {
	if len(items) == 0 {
		return "", false
	}
	total := 0.0
	for _, it := range items {
		total += it.weight
	}
	r := rng.Float64() * total
	acc := 0.0
	for _, it := range items {
		acc += it.weight
		if r < acc {
			return it.key, true
		}
	}
	return items[len(items)-1].key, true
}

func (p *Phonotactics) draw(table map[string]int, rng Rng, prev string) ([]string, bool) {
	if len(table) == 0 {
		return nil, true
	}
	keys := make([]string, 0, len(table))
	for k := range table {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool { return lessCluster(keys[i], keys[j]) })
	items := make([]weightedItem, 0, len(keys))
	A := float64(p.AlphabetSize())
	for _, k := range keys {
		count := float64(table[k])
		parts := clusterParts(k)
		weight := count
		if len(parts) > 0 {
			weight = count * (p.Prob(prev, parts[0]) * A)
		}
		items = append(items, weightedItem{k, weight})
	}
	chosen, ok := sample(items, rng)
	if !ok {
		return nil, false
	}
	return clusterParts(chosen), true
}

func (p *Phonotactics) drawNucleus(rng Rng, prev string, stress int) string {
	keys := make([]string, 0, len(p.Nuclei))
	for k := range p.Nuclei {
		if IsVowel(k) {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)
	items := make([]weightedItem, 0, len(keys))
	A := float64(p.AlphabetSize())
	for _, k := range keys {
		items = append(items, weightedItem{k, float64(p.Nuclei[k]) * p.Prob(prev, k) * A})
	}
	chosen, ok := sample(items, rng)
	if !ok {
		chosen = "AH0"
	}
	return WithStress(chosen, stress)
}

func (p *Phonotactics) sampleSyllables(rng Rng, lo, hi int) int {
	var ns []int
	for n := range p.SyllableCounts {
		if lo <= n && n <= hi {
			ns = append(ns, n)
		}
	}
	sort.Ints(ns)
	items := make([]weightedItem, 0, len(ns))
	for _, n := range ns {
		items = append(items, weightedItem{fmt.Sprint(n), float64(p.SyllableCounts[n])})
	}
	chosen, ok := sample(items, rng)
	if !ok {
		return lo
	}
	n := 0
	fmt.Sscan(chosen, &n)
	return n
}

// Build coins a pronounceable word from the habits (see phonotactics.py's build); syllables 0 means
// "as many as words usually have, between min and max"; avoid lists pronunciations not to return.
func (p *Phonotactics) Build(rng Rng, syllables, minSyllables, maxSyllables int, avoid []Entry, tries int) []string {
	if p.Words == 0 {
		panic("fit the phonotactics on some pronunciations first")
	}
	forbidden := map[string]bool{}
	for _, e := range avoid {
		forbidden[strings.Join(StripStress(e.Phones), " ")] = true
	}
	for try := 0; try < tries; try++ {
		n := syllables
		if n == 0 {
			n = p.sampleSyllables(rng, minSyllables, maxSyllables)
		}
		stressed := 0
		if n != 1 {
			if rng.Float64() < 0.6 {
				stressed = 0
			} else {
				stressed = rng.RandBelow(n)
			}
		}
		var phones []string
		ok := true
		for i := 0; i < n; i++ {
			prev := Boundary
			if len(phones) > 0 {
				prev = phones[len(phones)-1]
			}
			table := p.OnsetsMedial
			if i == 0 {
				table = p.OnsetsInitial
			}
			onset, good := p.draw(table, rng, prev)
			if !good {
				ok = false
				break
			}
			phones = append(phones, onset...)
			prev = Boundary
			if len(phones) > 0 {
				prev = phones[len(phones)-1]
			}
			stress := 0
			if i == stressed {
				stress = 1
			}
			nucleus := p.drawNucleus(rng, prev, stress)
			phones = append(phones, nucleus)
			table = p.CodasMedial
			if i == n-1 {
				table = p.CodasFinal
			}
			coda, good := p.draw(table, rng, nucleus)
			if !good {
				ok = false
				break
			}
			phones = append(phones, coda...)
		}
		if !ok || !WellFormed(phones) {
			continue
		}
		if forbidden[strings.Join(StripStress(phones), " ")] {
			continue
		}
		return phones
	}
	return nil
}

// Blend is the best portmanteau of two words: the phones, the syllables kept of a and dropped of b.
func (p *Phonotactics) Blend(a, b []string) ([]string, int, int, error) {
	hasVowel := func(ph []string) bool {
		for _, x := range ph {
			if IsVowel(x) {
				return true
			}
		}
		return false
	}
	if !hasVowel(a) || !hasVowel(b) {
		return nil, 0, 0, fmt.Errorf("both words need a vowel to be blended")
	}
	sa, sb := Syllabify(a, true), Syllabify(b, true)
	type candidate struct {
		phones []string
		i, j   int
	}
	var candidates []candidate
	flat := func(s []Syllable) []string {
		var out []string
		for _, syl := range s {
			out = append(out, syl.Phones()...)
		}
		return out
	}
	for i := 1; i <= len(sa); i++ {
		for j := 0; j < len(sb); j++ {
			head := flat(sa[:i])
			tail := flat(sb[j:])
			candidates = append(candidates, candidate{append(append([]string(nil), head...), tail...), i, j})
			headOpen := append(flat(sa[:i-1]), sa[i-1].Onset...)
			rest := append(append([]string(nil), sb[j].Rime()...), flat(sb[j+1:])...)
			candidates = append(candidates, candidate{append(append([]string(nil), headOpen...), rest...), i, j})
		}
	}
	bestTotal := 0.0
	var best *candidate
	longest := float64(len(a))
	if float64(len(b)) > longest {
		longest = float64(len(b))
	}
	for idx := range candidates {
		c := candidates[idx]
		if !WellFormed(c.phones) || len(c.phones) < 2 {
			continue
		}
		joint := p.Score(c.phones)
		balance := -math.Abs(float64(c.i)/float64(len(sa))-0.5) - math.Abs(float64(len(sb)-c.j)/float64(len(sb))-0.5)
		length := float64(len(c.phones)) / longest
		total := joint + 0.5*balance + 0.2*length
		if best == nil || total > bestTotal {
			best, bestTotal = &candidates[idx], total
		}
	}
	if best == nil {
		return append(append([]string(nil), a...), b...), len(sa), 0, nil
	}
	phones := append([]string(nil), best.phones...)
	primaries := 0
	for _, ph := range phones {
		if strings.HasSuffix(ph, "1") {
			primaries++
		}
	}
	if primaries > 1 {
		seen := false
		for k, ph := range phones {
			if strings.HasSuffix(ph, "1") {
				if seen {
					phones[k] = WithStress(ph, 2)
				}
				seen = true
			}
		}
	}
	if primaries == 0 {
		for k, ph := range phones {
			if IsVowel(ph) {
				phones[k] = WithStress(ph, 1)
				break
			}
		}
	}
	return phones, best.i, best.j, nil
}

// String is the one-line description.
func (p *Phonotactics) String() string {
	return fmt.Sprintf("Phonotactics(words=%d, pairs=%d, stress=%t)", p.Words, len(p.Bigrams), p.Stress)
}
